code/datums/status_effects/debuffs/fire_stacks.dm

```diff
@@ -0,0 +1,376 @@
+
+/**
+ * Used to update owner's effect overlay
+ */
+
+/datum/status_effect/fire_handler/proc/update_overlay()
+
+/datum/status_effect/fire_handler/fire_stacks
+	id = "fire_stacks" //fire_stacks and wet_stacks should have different IDs or else has_status_effect won't work
+
+	enemy_types = list(/datum/status_effect/fire_handler/wet_stacks)
+	stack_modifier = 1
+
+	/// If we're on fire
+	var/on_fire = FALSE
+	/// A weakref to the mob light emitter
+	var/datum/weakref/firelight_ref
+	/// Type of mob light emitter we use when on fire
+	var/firelight_type = /obj/effect/dummy/lighting_obj/moblight/fire
+	/// Stores current fire overlay icon state, for optimisation purposes
+	var/last_icon_state
+
+/datum/status_effect/fire_handler/fire_stacks/tick(delta_time, times_fired)
+	if(stacks <= 0)
+		qdel(src)
+		return TRUE
+
+	if(!on_fire || isanimal(owner))
+		return TRUE
+
+	adjust_stacks(-0.05 * delta_time)
+
+	if(iscyborg(owner))
+		adjust_stacks(-0.5 * delta_time)
+
+	if(stacks <= 0)
+		qdel(src)
+		return TRUE
+
+	var/datum/gas_mixture/air = owner.loc.return_air()
+	if(!air.gases[/datum/gas/oxygen] || air.gases[/datum/gas/oxygen][MOLES] < 1)
+		qdel(src)
+		return TRUE
+
+	deal_damage(delta_time, times_fired)
+	update_overlay()
+
+/**
+ * Proc that handles damage dealing and all special effects
+ *
+ * Arguments:
+ * - delta_time
+ * - times_fired
+ *
+ */
+
+/datum/status_effect/fire_handler/proc/deal_damage(delta_time, times_fired)
+	if(ishuman(owner))
+		var/mob/living/carbon/human/victim = owner
+		SEND_SIGNAL(victim, COMSIG_HUMAN_BURNING)
+		burn_stuff(delta_time, times_fired)
+		var/no_protection = FALSE
+		if(victim.dna && victim.dna.species)
+			no_protection = victim.dna.species.handle_fire(victim, delta_time, times_fired, no_protection)
+		harm_human(delta_time, times_fired, no_protection)
+
+	if(isalien(owner))
+		owner.adjust_bodytemperature(BODYTEMP_HEATING_MAX * 0.5 * delta_time)
+
+	var/turf/location = get_turf(owner)
+	location.hotspot_expose(700, 25 * delta_time, TRUE)
+
+/**
+ * Used to deal damage to humans and count their protection.
+ *
+ * Arguments:
+ * - delta_time
+ * - times_fired
+ * - no_protection: When set to TRUE, fire will ignore any possible fire protection
+ *
+ */
+
+/datum/status_effect/fire_handler/proc/harm_human(delta_time, times_fired, no_protection = FALSE)
+	var/mob/living/carbon/human/victim = owner
+	var/thermal_protection = victim.get_thermal_protection()
+
+	if(thermal_protection >= FIRE_IMMUNITY_MAX_TEMP_PROTECT && !no_protection)
+		return
+
+	if(thermal_protection >= FIRE_SUIT_MAX_TEMP_PROTECT && !no_protection)
+		victim.adjust_bodytemperature(5.5 * delta_time)
+	else
+		victim.adjust_bodytemperature((BODYTEMP_HEATING_MAX + (stacks * 12)) * 0.5 * delta_time)
+		SEND_SIGNAL(victim, COMSIG_ADD_MOOD_EVENT, "on_fire", /datum/mood_event/on_fire)
+		victim.mind?.add_memory(MEMORY_FIRE, list(DETAIL_PROTAGONIST = victim), story_value = STORY_VALUE_OKAY)
```

Comment:
```suggestion
	if(no_protection)
		victim.adjust_bodytemperature((BODYTEMP_HEATING_MAX + (stacks * 12)) * 0.5 * delta_time)
		SEND_SIGNAL(victim, COMSIG_ADD_MOOD_EVENT, "on_fire", /datum/mood_event/on_fire)
		victim.mind?.add_memory(MEMORY_FIRE, list(DETAIL_PROTAGONIST = victim), story_value = STORY_VALUE_OKAY)
		return

	if(thermal_protection >= FIRE_IMMUNITY_MAX_TEMP_PROTECT)
		return
	if(thermal_protection >= FIRE_SUIT_MAX_TEMP_PROTECT)
		victim.adjust_bodytemperature(5.5 * delta_time)
```
Also for the record, I'd prefer no_protection be called protection. Negatives and double negatives can be hard to parse, best to just be as direct as possible to start, and then let the ! operator do the talking